### PR TITLE
Replace references to _DataLoaderIter with _BaseDataLoaderIter

### PR DESCRIPTION
**Pull request by ngoldbaum**
Back in April, @malmaud added type annotations for `dataloader.py`. However, at about the same time, @Ssnl in #19228 replaced `_DataLoaderIter` with `_BaseDataLoaderIter` and two subclasses, `_SingleProcessDataLoaderIter`, and `_MultiProcessingDataLoaderIter`. However - probably because these changes happened in parallel at roughly the same time, the type stubs and several other references in the codebase were never updated to match this refactoring.

I've gone ahead and done the updates to reflect the refactoring in #19228, which fixes the specific type stub/impelementation mismatch pointed out in #26673, although not the broader problem that pytorch doesn't have a test to make sure that the `.pyi` type stub files match the real API defined in `.py` files.

### CODE REVIEW

File: torch/utils/data/dataloader.pyi [hunk 1]
```diff
@@ -28,12 +28,10 @@ class DataLoader(Generic[T_co]):
                  worker_init_fn: _worker_init_fn_t=...) -> None: ...
 
     def __len__(self) -> int: ...
-    # We quote '_DataLoaderIter' since it isn't defined yet and the definition can't be moved up since
-    # '_DataLoaderIter' references 'DataLoader'. Pending updates of PEP 484 will fix this.
-    def __iter__(self) -> '_DataLoaderIter':...
+    def __iter__(self) -> _BaseDataLoaderIter:...
```
Comment (SsnL) [inline torch/utils/data/dataloader.pyi]: Has the problem mentioned in previous comment been solved?

**Review by ssnl — COMMENTED** [review]
Thank you for doing this!

File: torch/utils/data/dataloader.pyi (hunk 1 above)
Comment (ngoldbaum) [reply torch/utils/data/dataloader.pyi]: I think so? I wasn’t able to reproduce the error in the comments with the latest version of mypy as well as a version from April, I was running:

```
mypy torch/utils/data/dataloader.py
```

and didn’t see any errors from mypy. That invocation may not have been sufficient to trigger the error @malmaud saw when they originally wrote this code.

Comment (malmaud) [reply torch/utils/data/dataloader.pyi]: At some point relatively recently, Python/mypy added the ability to handle unquoted forward references, but there are definitely versions which cannot. I think it's safer to keep the quotes unless I was actually in error before and this isn't a forward reference.

Comment (ezyang) [reply torch/utils/data/dataloader.pyi]: @ngoldbaum mind updating this? :)

Comment (ngoldbaum) [reply torch/utils/data/dataloader.pyi]: Done, I also expanded the comment to reflect the current situation in mypy so future readers are hopefully a little less confused.

**Review by ezyang — APPROVED** [review]
waiting on ci

**Pull request closed without merging**